sdk/identity/Azure.Identity/src/TokenCache.cs

```diff
@@ -0,0 +1,289 @@
+
+        private async Task OnBeforeCacheAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                args.TokenCache.DeserializeMsalV3(_data, true);
+
+                _cacheAccessMap.GetOrCreateValue(args.TokenCache).Update();
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+
+        private async Task OnAfterCacheAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            if (args.HasStateChanged)
+            {
+                await UpdateCacheDataAsync(args.TokenCache).ConfigureAwait(false);
+            }
+        }
+
+        private async Task UpdateCacheDataAsync(ITokenCacheSerializer tokenCache)
+        {
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                if (!_cacheAccessMap.TryGetValue(tokenCache, out CacheTimestamp lastRead) || lastRead.Value < _lastUpdated)
+                {
+                    _data = await MergeCacheData(_data, tokenCache.SerializeMsalV3()).ConfigureAwait(false);
+                }
+                else
+                {
+                    _data = tokenCache.SerializeMsalV3();
+                }
+
+                _cacheAccessMap.GetOrCreateValue(tokenCache).Update();
+
+                _lastUpdated = DateTime.UtcNow;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            if (Updated != null)
+            {
+                await Updated.Invoke(new TokenCacheUpdatedArgs(this)).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<byte[]> MergeCacheData(byte[] cacheA, byte[] cacheB)
+        {
+            byte[] merged = null;
+
+            var client = PublicClientApplicationBuilder.Create(Guid.NewGuid().ToString()).Build();
+
+            client.UserTokenCache.SetBeforeAccess(args => args.TokenCache.DeserializeMsalV3(cacheA));
+
+            await client.GetAccountsAsync().ConfigureAwait(false);
+
+            client.UserTokenCache.SetBeforeAccess(args => args.TokenCache.DeserializeMsalV3(cacheB, shouldClearExistingCache: false));
+
+            client.UserTokenCache.SetAfterAccess(args => merged = args.TokenCache.SerializeMsalV3());
+
+            await client.GetAccountsAsync().ConfigureAwait(false);
+
+            return merged;
+        }
+
+        /// <summary>
+        /// Disposes of the <see cref="TokenCache"/>.
+        /// </summary>
+        /// <param name="disposing">Indicates whether managed resources should be disposed.</param>
+        protected virtual void Dispose(bool disposing)
```

Comment:
Do we define the Dispose just for the `_lock`? IIRC it doesn't need to be disposed if we are not creating the  WaitHandle.

Reply:
Do we have control over when the `SemaphoreSlim` will need to create the underlying WaitHandle?

Reply:
Only when you call `AvailableWaitHandle`

Reply:
Perhaps we can remove the need for this class to be IDisposable. As I called out below I'd rather defer changing it here  so the API is consistent with the released 1.4.0-beta.1 API. I've filed issue https://github.com/Azure/azure-sdk-for-net/issues/17124 to investigate and remove if unneeded.